src/modules/instant_production/module.ts

```diff
@@ -60,24 +60,42 @@ function moveActionMixin(
       super.process(action);
 
       const grid = this.grid();
-      if (hasStartingTown(grid, action)) {
-        // Automatically apply instant production since there is no choice
+      const start = grid.get(action.startingCity);
+      const end = grid.get(action.path[action.path.length - 1].endingStop);
+
+      const startCanInstantProduction = isInstantProductionTarget(start);
+      const endCanInstantProduction = isInstantProductionTarget(end);
+
+      // If neither side can instant production, just end the turn; no instant production
```

Comment:
Seems like you should be able to use a filter here.

If filtered.size == 0 end turn
If filtered.size == 1 apply filtered[0]
Else choose from filtered.

Reply:
Nice, that's much better. TS/JS isn't my daily-driver language so I tend to miss these idioms. :)